From 85216f2db006a5f51703c0bcd519e9d630e9b7eb Mon Sep 17 00:00:00 2001
From: Eduard Tolosa <tolosaeduard@gmail.com>
Date: Tue, 23 Sep 2025 20:08:58 +0000
Subject: [PATCH] clutter/gesture: Do not crash on unknown events

The current gestures logic depends on knowing the event type before
handling it, and resorts to `g_assert_not_reached()` in the case of
unknown/null events. That does cause issues like
https://gitlab.gnome.org/GNOME/mutter/-/issues/4340.

Even with https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/4676
merged, which handles the null events, hypothetical new event types
will still make gnome-shell crash. Let's just bail in that case.

For reference, this was discussed on
https://gitlab.gnome.org/GNOME/mutter/-/issues/4340#note_2556912,
https://gitlab.gnome.org/GNOME/mutter/-/issues/4340#note_2557144

clutter/gesture: Bail early for unknown gesture events

clutter/gesture: Only allow defined gesture events

clutter/gesture: Fix code style for new code

Part-of: <https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/4679>
---
 clutter/clutter/clutter-gesture.c | 46 ++++++++++++++++---------------
 1 file changed, 24 insertions(+), 22 deletions(-)

diff --git a/clutter/clutter/clutter-gesture.c b/clutter/clutter/clutter-gesture.c
index 8a058889742..6c1c43a9343 100644
--- a/clutter/clutter/clutter-gesture.c
+++ b/clutter/clutter/clutter-gesture.c
@@ -817,6 +817,27 @@ is_sequence_end_event (const ClutterEvent *event)
     }
 }
 
+static gboolean
+supported_gesture_event (ClutterEventType event_type)
+{
+  switch (event_type)
+    {
+    case CLUTTER_ENTER:
+    case CLUTTER_LEAVE:
+    case CLUTTER_BUTTON_PRESS:
+    case CLUTTER_MOTION:
+    case CLUTTER_BUTTON_RELEASE:
+    case CLUTTER_TOUCH_BEGIN:
+    case CLUTTER_TOUCH_UPDATE:
+    case CLUTTER_TOUCH_END:
+    case CLUTTER_TOUCH_CANCEL:
+      return TRUE;
+
+    default:
+      return FALSE;
+    }
+}
+
 static gboolean
 clutter_gesture_handle_event (ClutterAction      *action,
                               const ClutterEvent *event)
@@ -838,13 +859,10 @@ clutter_gesture_handle_event (ClutterAction      *action,
   gboolean may_remove_point = TRUE;
   ClutterGestureState old_state = priv->state;
 
-  if (event_type == CLUTTER_TOUCHPAD_PINCH ||
-      event_type == CLUTTER_TOUCHPAD_SWIPE ||
-      event_type == CLUTTER_TOUCHPAD_HOLD ||
-      event_type == CLUTTER_SCROLL)
+  if (clutter_event_get_flags (event) & CLUTTER_EVENT_FLAG_SYNTHETIC)
     return CLUTTER_EVENT_PROPAGATE;
 
-  if (clutter_event_get_flags (event) & CLUTTER_EVENT_FLAG_SYNTHETIC)
+  if (!supported_gesture_event (event_type))
     return CLUTTER_EVENT_PROPAGATE;
 
   sprite = clutter_backend_get_sprite (backend, stage, event);
@@ -927,25 +945,9 @@ clutter_gesture_handle_event (ClutterAction      *action,
       seq_data->latest_event = clutter_event_copy (event);
 
       priv->latest_index = seq_index;
-
       seq_data->seen = TRUE;
 
-      switch (event_type)
-        {
-        case CLUTTER_BUTTON_PRESS:
-        case CLUTTER_MOTION:
-        case CLUTTER_BUTTON_RELEASE:
-        case CLUTTER_TOUCH_BEGIN:
-        case CLUTTER_TOUCH_UPDATE:
-        case CLUTTER_TOUCH_END:
-        case CLUTTER_TOUCH_CANCEL:
-          handle_pointer_event (self, seq_index, event);
-          break;
-
-        default:
-          g_assert_not_reached ();
-          break;
-        }
+      handle_pointer_event (self, seq_index, event);
     }
 
   if (may_remove_point && is_sequence_end_event (event))
-- 
GitLab

